openssl-dynamic/src/main/c/ssl.c

```diff
@@ -1777,6 +1777,141 @@ TCN_IMPLEMENT_CALL(void, SSL, freeX509Chain)(TCN_STDARGS, jlong x509Chain)
     sk_X509_pop_free(chain, X509_free);
 }
 
+TCN_IMPLEMENT_CALL(void, SSL, setKeyMaterialServerSide)(TCN_STDARGS, jlong ssl, jlong chain, jlong key)
+{
+#if OPENSSL_VERSION_NUMBER < 0x10002000L || defined(LIBRESSL_VERSION_NUMBER)
+    // Only supported on openssl 1.0.2+
+    tcn_Throw(e, "Not supported");
+#else
+    SSL *ssl_ = J2P(ssl, SSL *);
+
+    TCN_CHECK_NULL(ssl_, ssl, /* void */);
+
+    EVP_PKEY* pkey = J2P(key, EVP_PKEY *);
+    STACK_OF(X509) *cchain = J2P(chain, STACK_OF(X509) *);
+    int numCerts;
+    char err[ERR_LEN];
+    int i;
+
+    UNREFERENCED(o);
+    TCN_ASSERT(ssl != NULL);
+
+    TCN_CHECK_NULL(cchain, chain, /* void */);
+
+    numCerts = sk_X509_num(cchain);
+
+    // SSL_use_certificate will increment the reference count of the cert.
+    if (numCerts <= 0 || SSL_use_certificate(ssl_, sk_X509_value(cchain, 0)) <= 0) {
+        ERR_error_string_n(ERR_get_error(), err, ERR_LEN);
+        ERR_clear_error();
+        tcn_Throw(e, "Error setting certificate (%s)", err);
+        return;
+    }
+
+    if (pkey != NULL) {
+        // SSL_use_PrivateKey will increment the reference count of the key.
+        if (SSL_use_PrivateKey(ssl_, pkey) <= 0) {
+            ERR_error_string_n(ERR_get_error(), err, ERR_LEN);
+            ERR_clear_error();
+            tcn_Throw(e, "Error setting private key (%s)", err);
+            return;
+        }
+        if (SSL_check_private_key(ssl_) <= 0) {
+            ERR_error_string_n(ERR_get_error(), err, ERR_LEN);
+            ERR_clear_error();
+
+            tcn_Throw(e, "Private key does not match the certificate public key (%s)", err);
+            return;
+        }
+    }
+
+    // The first cert was loaded via SSL_use_certificate so skip it.
+    for (i = 1; i < numCerts; ++i) {
+
+        // SSL_add1_chain_cert will increment the reference count of the cert.
+        if (SSL_add1_chain_cert(ssl_, sk_X509_value(cchain, i)) != 1) {
```

Comment:
dumb question: how does this unref if there is an error?

Reply:
if there is an error the reference count of the cert is not incremented. All the previous added certificates will be part of the SSL lifetime which means once its teared down the reference count of these will be decremented and so eventually freed.